rslib/src/scheduler/queue/builder/mod.rs

```diff
@@ -204,134 +210,132 @@ fn sort_learning(mut learning: Vec<DueCard>) -> VecDeque<LearningQueueEntry> {
+    fn add_new_cards_by_deck(&self, queues: &mut QueueBuilder, ctx: &mut Context) -> Result<()> {
+        // TODO: must own Vec as closure below requires unique access to ctx
+        // maybe decks should not be field of Context?
+        for deck_id in ctx.limits.remaining_decks() {
+            if ctx.limits.is_exhausted_root() {
                 break;
             }
-            let limit = remaining.get_mut(&deck.id).unwrap();
-            if limit.new > 0 {
-                self.storage
-                    .for_each_new_card_in_deck(deck.id, reverse, |card| {
-                        let bury = get_bury_mode(card.original_deck_id.or(deck.id));
-                        if limit.new != 0 {
-                            if queues.add_new_card(card, bury) {
-                                limit.new -= 1;
-                                selected_deck_limits.new =
-                                    selected_deck_limits.new.saturating_sub(1);
-                            }
-
-                            true
-                        } else {
-                            false
+            if !ctx.limits.is_exhausted(deck_id) {
+                self.storage.for_each_new_card_in_deck(deck_id, |card| {
+                    let bury = ctx.bury_mode(card.original_deck_id.or(deck_id));
+                    // TODO: This could be done more efficiently if we held on to the node_id
+                    // and only adjusted the parent nodes after this node's limit is reached
+                    if let Some(node_id) = ctx.limits.remaining_node_id(deck_id) {
+                        if queues.add_new_card(card, bury) {
+                            ctx.limits.decrement_node_and_parent_limits(&node_id, true);
                         }
-                    })?;
+
+                        true
+                    } else {
+                        false
+                    }
+                })?;
             }
         }
 
-        let final_limits = RemainingLimits {
-            new: selected_deck_limits_at_start
-                .new
-                .min(selected_deck_limits.review),
-            ..selected_deck_limits_at_start
-        };
-        let queues = queues.build(
-            final_limits,
-            self.learn_ahead_secs() as i64,
-            timing.days_elapsed,
-        );
+        Ok(())
+    }
 
-        Ok(queues)
+    fn add_new_cards_by_position(
+        &self,
+        queues: &mut QueueBuilder,
+        ctx: &mut Context,
+        reverse: bool,
+    ) -> Result<()> {
+        self.storage
+            .for_each_new_card_in_active_decks(reverse, |card| {
+                let bury = ctx.bury_mode(card.original_deck_id.or(card.current_deck_id));
+                if let Some(node_id) = ctx.limits.remaining_node_id(card.current_deck_id) {
+                    if queues.add_new_card(card, bury) {
+                        ctx.limits.decrement_node_and_parent_limits(&node_id, true);
```

Comment:
This code looks to be exiting early when limits have been reached.  That won't work with the approach that the code currently uses, which is to gather max(deck, ...parent decks) from each deck, sort the received cards, and then truncate them to the global limit.

Reply:
Just stumbled across the same thing. :)